partner_mobile_duplicate_warn/models/res_partner.py

```diff
@@ -15,26 +16,76 @@ class ResPartner(models.Model):
         compute_sudo=True,
     )
 
-    @api.depends("mobile", "company_id")
+    @api.depends(lambda x: x._get_same_mobile_depends())
     def _compute_same_mobile_partner_id(self):
+        empty_recordset = self.env[self._name]
+        for partner in self:
+            partner.same_mobile_partner_id = (
+                partner._find_same_mobile_partner()
+                if partner.mobile
+                else empty_recordset
+            )
+
+    def _find_same_mobile_partner(self):
+        """Find one partner with the same mobile."""
+        self.ensure_one()
+        domain = self._get_same_mobile_domain()
+        return self.with_context(active_test=False).search(domain, limit=1)
+
+    @api.model
+    def _get_same_mobile_depends(self):
+        """Return the fields on which same_mobile_partner_id depends.
+
+        Return the fields used in _get_same_mobile_domain function.
+        """
+        return ["mobile", "company_id"]
+
+    def _get_same_mobile_domain(self, exclude_self=True):
+        """Return domain to find partners with same mobile.
+
+        By default we want to find other partners with the
+        same mobile, but if we want to see if both self and
+        another partner have the same mobile within a domain,
+        we will leave out the domain leave that excludes self.
+
+        This can be used to prevent false positives when we want a
+        constraint forbidding duplicates.
+        """
+        self.ensure_one()
         # With phone_validation, the "mobile" field should be
         # clean in E.164 format, without any start/ending spaces
         # So we search on the 'mobile' field with '=' !
-        for partner in self:
-            same_mobile_partner_id = False
-            if partner.mobile:
-                domain = [("mobile", "=", partner.mobile)]
-                if partner.company_id:
-                    domain += [
-                        "|",
-                        ("company_id", "=", False),
-                        ("company_id", "=", partner.company_id.id),
-                    ]
-                partner_id = partner._origin.id
-                if partner_id:
-                    domain.append(("id", "!=", partner_id))
-                same_mobile_partner = self.with_context(active_test=False).search(
-                    domain, limit=1
-                )
-                same_mobile_partner_id = same_mobile_partner.id or False
-            partner.same_mobile_partner_id = same_mobile_partner_id
+        domain = [("mobile", "=", self.mobile)]
+        if self.company_id:
+            domain += [
+                "|",
+                ("company_id", "=", False),
+                ("company_id", "=", self.company_id.id),
+            ]
+        self_id = (self._origin.id or self.id) if exclude_self else False
```

Comment:
Shouldn't this be the other way around?

Reply:
> Shouldn't this be the other way around?

nope

Reply:
I went from include to exclude to get rid of a double negation.